src/protocols/secure_channel/CASESession.cpp

```diff
@@ -129,6 +132,164 @@ static constexpr ExchangeContext::Timeout kExpectedLowProcessingTime    = System
 static constexpr ExchangeContext::Timeout kExpectedSigma1ProcessingTime = kExpectedLowProcessingTime;
 static constexpr ExchangeContext::Timeout kExpectedHighProcessingTime   = System::Clock::Seconds16(30);
 
+// Helper for managing a session's outstanding work.
+// Holds work data which is provided to a scheduled work callback (standalone),
+// then (if not canceled) to a scheduled after work callback (on the session).
+template <class DATA>
+class CASESession::WorkHelper
+{
+public:
+    // Work callback, processed in the background via `PlatformManager::ScheduleBackgroundWork`.
+    // This is a non-member function which does not use the associated session.
+    // The return value is passed to the after work callback (called afterward).
+    // Set `cancel` to true if calling the after work callback is not necessary.
+    typedef CHIP_ERROR (*WorkCallback)(DATA & data, bool & cancel);
+
+    // After work callback, processed in the main CHIP task via `PlatformManager::ScheduleWork`.
+    // This is a member function to be called on the associated session after the work callback.
+    // The `status` value is the result of the work callback (called beforehand).
+    typedef CHIP_ERROR (CASESession::*AfterWorkCallback)(DATA & data, CHIP_ERROR status);
+
+    // Create a work helper using the specified session, work callback, after work callback, and data (template arg).
+    // Lifetime is not managed, see `Create` for that option.
+    WorkHelper(CASESession & session, WorkCallback workCallback, AfterWorkCallback afterWorkCallback) :
+        mSession(&session), mWorkCallback(workCallback), mAfterWorkCallback(afterWorkCallback)
+    {}
+
+    // Create a work helper using the specified session, work callback, after work callback, and data (template arg).
+    // Lifetime is managed by sharing between the caller (typically the session) and the helper itself (while work is scheduled).
+    static std::shared_ptr<WorkHelper> Create(CASESession & session, WorkCallback workCallback, AfterWorkCallback afterWorkCallback)
+    {
+        std::shared_ptr<WorkHelper> ptr(Platform::New<WorkHelper>(session, workCallback, afterWorkCallback),
+                                        Platform::Delete<WorkHelper>);
```

Comment:
This construct is centrally available here:

https://github.com/project-chip/connectedhomeip/blob/55a377b73597cee11fdce176aa9a0132f5a13108/src/lib/support/CHIPMem.h#L191

However, after introducing this, it was pointed out to me by @mrjerryjohns that doing this still bypasses the `chip::Platform` allocator for storing the `shared_ptr` control block.

I am not personally opposed to this approach based upon this, but think it is worth mentioning.

Reply:
OK I will look into this one. Yes, the control block is another allocation, I believe C++20 has some stuff for customizing it, but we aren't using that yet.